File: web/frontend/assets/js/papers-catalog.js
(function () {
  'use strict';

  const ASSET_VERSION = '20260714n2';
  const MANIFEST_URL = '/assets/data/papers-manifest.json?v=20260714n2';
  const SCHEMA_VERSION = 'papers-manifest-v2';
  const SLUG_PATTERN = '^(?!.*\\.\\.)(?!.*\\.$)[a-z0-9][a-z0-9._-]{1,119}$';
  const SLUG_RE = new RegExp(SLUG_PATTERN);
  const SOURCE_ROOT = '/dada8899/structural-isomorphism/';
  const GROUP_ORDER = ['unified', 'arxiv-drafts', 'phase-papers', 'tutorials'];
  const GROUP_COUNTS = Object.freeze({
    unified: 1,
    'arxiv-drafts': 4,
    'phase-papers': 14,
    tutorials: 1,
  });
  const STATUS_COUNTS = Object.freeze({
    'historical-record': 14,
    'historical-draft': 5,
    'historical-tutorial': 1,
  });
  const STATUS_LABELS = Object.freeze({
    'historical-record': Object.freeze({ zh: '历史结果记录', en: 'Historical result record' }),
    'historical-draft': Object.freeze({ zh: '历史研究稿', en: 'Historical research draft' }),
    'historical-tutorial': Object.freeze({ zh: '历史复现教程', en: 'Historical reproduction tutorial' }),
  });
  const REQUIRED_CONTRACT = Object.freeze({
    schema_version: 'empirical-result-card-v1',
    evidence_level: 'historical_internal_record',
    outcome_status: 'not_normalized_in_current_ledger',
    source_status: 'inspect_record',
    license_status: 'not_recorded_in_manifest',
    preregistration_status: 'not_recorded_in_manifest',
    ledger_status: 'not_bound',
    review_status: 'internal_only',
  });
  const CONTROL_RE = /[\u0000-\u0008\u000b\u000c\u000e-\u001f\u007f]/;
  let manifestPromise = null;

  function own(object, key) {
    return Object.prototype.hasOwnProperty.call(object, key);
  }

  function requiredText(value, field, maxLength) {
    if (
      typeof value !== 'string' || !value.trim() || value.length > maxLength ||
      CONTROL_RE.test(value)
    ) {
      throw new Error(`Invalid papers manifest field: ${field}`);
    }
    return value.trim();
  }

  function optionalText(value, field, maxLength) {
    if (value == null || value === '') return '';
    return requiredText(value, field, maxLength);
  }

  function integer(value, field, minimum) {
    if (!Number.isInteger(value) || value < minimum) {
      throw new Error(`Invalid papers manifest field: ${field}`);
    }
    return value;
  }

  function validateSlug(value) {
    const slug = requiredText(value, 'paper.slug', 120);
    if (!SLUG_RE.test(slug)) throw new Error(`Invalid paper slug: ${slug}`);
    return slug;
  }

  function validateSourceUrl(value) {
    const source = requiredText(value, 'paper.source_url', 500);
    let parsed;
    try {
      parsed = new URL(source);
    } catch (_error) {
      throw new Error('Invalid paper source URL');
    }
    if (
      parsed.protocol !== 'https:' || parsed.hostname !== 'github.com' || parsed.port ||
      parsed.username || parsed.password || parsed.search || parsed.hash ||
      !parsed.pathname.startsWith(SOURCE_ROOT)
    ) {
      throw new Error('Invalid paper source URL');
    }
    return parsed.toString();
  }

  function validateContract(value) {
    if (!value || typeof value !== 'object' || Array.isArray(value)) {
      throw new Error('Missing empirical result-card contract');
    }
    Object.keys(REQUIRED_CONTRACT).forEach((key) => {
      if (value[key] !== REQUIRED_CONTRACT[key]) {
        throw new Error(`Invalid empirical result-card contract: ${key}`);
      }
    });
    [
      'observed_fallback_zh', 'observed_fallback_en', 'boundary_zh',
      'boundary_en', 'next_test_zh', 'next_test_en',
    ].forEach((key) => requiredText(value[key], `result_contract.${key}`, 900));
    return value;
  }

  function validatePaper(value, seen, statusCounts) {
    if (!value || typeof value !== 'object' || Array.isArray(value)) {
      throw new Error('Invalid paper record');
    }
    const slug = validateSlug(value.slug);
    if (seen.has(slug)) throw new Error(`Duplicate paper slug: ${slug}`);
    seen.add(slug);
    if (own(value, 'external_link')) throw new Error('Paper records must use internal detail routes');
    requiredText(value.title_zh, 'paper.title_zh', 300);
    requiredText(value.title_en, 'paper.title_en', 300);
    optionalText(value.class, 'paper.class', 240);
    optionalText(value.alpha, 'paper.alpha', 240);
    requiredText(value.date, 'paper.date', 10);
    if (!/^\d{4}-\d{2}-\d{2}$/.test(value.date)) throw new Error('Invalid paper date');
    integer(value.words, 'paper.words', 1);
    integer(value.minutes, 'paper.minutes', 1);
    if (value.n_tail != null) integer(value.n_tail, 'paper.n_tail', 1);
    if (!own(STATUS_LABELS, value.status)) throw new Error('Invalid paper status');
    validateSourceUrl(value.source_url);
    statusCounts[value.status] += 1;
    return value;
  }

  function validateManifest(value) {
    if (
      !value || typeof value !== 'object' || Array.isArray(value) ||
      !value.meta || typeof value.meta !== 'object' || !Array.isArray(value.groups)
    ) {
      throw new Error('Invalid papers manifest');
    }
    const meta = value.meta;
    if (meta.schema_version !== SCHEMA_VERSION) throw new Error('Unsupported papers manifest schema');
    if (meta.asset_version !== ASSET_VERSION) throw new Error('Papers asset version drift');
    if (meta.slug_pattern !== SLUG_PATTERN) throw new Error('Papers slug contract drift');
    const contract = validateContract(meta.result_contract);
    const seenGroups = new Set();
    const seenPapers = new Set();
    const statusCounts = {
      'historical-record': 0,
      'historical-draft': 0,
      'historical-tutorial': 0,
    };
    const records = [];

    value.groups.forEach((group, groupIndex) => {
      if (!group || typeof group !== 'object' || Array.isArray(group)) {
        throw new Error('Invalid paper group');
      }
      if (group.id !== GROUP_ORDER[groupIndex] || seenGroups.has(group.id)) {
        throw new Error('Invalid paper group order or duplicate group');
      }
      seenGroups.add(group.id);
      requiredText(group.title_zh, 'group.title_zh', 240);
      requiredText(group.title_en, 'group.title_en', 240);
      requiredText(group.desc_zh, 'group.desc_zh', 800);
      requiredText(group.desc_en, 'group.desc_en', 800);
      if (!Array.isArray(group.papers) || group.papers.length !== GROUP_COUNTS[group.id]) {
        throw new Error(`Paper group count drift: ${group.id}`);
      }
      group.papers.forEach((paper) => records.push(validatePaper(paper, seenPapers, statusCounts)));
    });

    if (seenGroups.size !== GROUP_ORDER.length) throw new Error('Missing paper group');
    const counts = {
      total: integer(meta.total_items, 'meta.total_items', 1),
      records: integer(meta.historical_result_records, 'meta.historical_result_records', 1),
      drafts: integer(meta.historical_research_drafts, 'meta.historical_research_drafts', 1),
      tutorials: integer(meta.historical_tutorials, 'meta.historical_tutorials', 1),
    };
    if (
      counts.total !== records.length || counts.total !== 20 ||
      counts.records !== STATUS_COUNTS['historical-record'] ||
      counts.drafts !== STATUS_COUNTS['historical-draft'] ||
      counts.tutorials !== STATUS_COUNTS['historical-tutorial'] ||
      counts.total !== counts.records + counts.drafts + counts.tutorials
    ) {
      throw new Error('Papers manifest count drift');
    }
    Object.keys(STATUS_COUNTS).forEach((status) => {
      if (statusCounts[status] !== STATUS_COUNTS[status]) {
        throw new Error(`Papers status count drift: ${status}`);
      }
    });
    const bySlug = Object.create(null);
    records.forEach((record) => { bySlug[record.slug] = record; });
    return { manifest: value, contract, records, bySlug, counts };
  }

  function escapeHtml(value) {
    const node = document.createElement('span');
    node.textContent = String(value == null ? '' : value);
    return node.innerHTML;
  }

  function locale() {
    try {
      return window.i18n && window.i18n.getLang && window.i18n.getLang() === 'en' ? 'en' : 'zh';
    } catch (_error) {
      return 'zh';
    }
  }

  function localized(record, field, selectedLocale) {
    const lang = selectedLocale === 'en' ? 'en' : 'zh';
    return record[`${field}_${lang}`] || record[`${field}_zh`] || record[`${field}_en`] || '';
  }

  function paperUrl(slug) {
    return `/paper/${encodeURIComponent(validateSlug(slug))}`;
  }

  function markdownUrl(slug) {
    return `/assets/data/papers/${encodeURIComponent(validateSlug(slug))}.md?v=${ASSET_VERSION}`;
  }

  function slugFromLocation(locationLike) {
    const pathname = String(locationLike && locationLike.pathname || '');
    const pathMatch = pathname.match(/^\/paper\/([^/]+)$/);
    let raw = pathMatch ? pathMatch[1] : '';
    if (!raw && pathname.endsWith('/paper.html')) {
      raw = new URLSearchParams(String(locationLike.search || '')).get('doc') || '';
    }
    try {
      raw = decodeURIComponent(raw);
    } catch (_error) {
      throw new Error('Invalid encoded paper slug');
    }
    return validateSlug(raw);
  }

  function loadManifest(options) {
    const force = Boolean(options && options.force);
    if (force) manifestPromise = null;
    if (!manifestPromise) {
      manifestPromise = fetch(MANIFEST_URL, {
        credentials: 'same-origin',
        cache: 'no-store',
        headers: { Accept: 'application/json' },
      }).then((response) => {
        if (!response.ok) throw new Error(`Manifest request failed (${response.status})`);
        return response.json();
      }).then(validateManifest).catch((error) => {
        manifestPromise = null;
        throw error;
      });
    }
    return manifestPromise;
  }

  window.StructuralPapersCatalog = Object.freeze({
    ASSET_VERSION,
    MANIFEST_URL,
    STATUS_LABELS,
    escapeHtml,
    locale,
    localized,
    loadManifest,
    markdownUrl,
    paperUrl,
    slugFromLocation,
    validateManifest,
    validateSlug,
    validateSourceUrl,
  });
})();

File: web/backend/services/papers_catalog.py
"""Fail-closed loader for the public historical-papers catalog."""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from functools import lru_cache
from pathlib import Path
from types import MappingProxyType
from typing import Any, Mapping
from urllib.parse import urlsplit


SCHEMA_VERSION = "papers-manifest-v2"
ASSET_VERSION_RE = re.compile(r"^[0-9]{8}[a-z][0-9]+$")
SLUG_PATTERN = r"^(?!.*\.\.)(?!.*\.$)[a-z0-9][a-z0-9._-]{1,119}$"
SLUG_RE = re.compile(SLUG_PATTERN)
SOURCE_PREFIX = "/dada8899/structural-isomorphism/"
GROUP_COUNTS = {
    "unified": 1,
    "arxiv-drafts": 4,
    "phase-papers": 14,
    "tutorials": 1,
}
STATUS_COUNTS = {
    "historical-record": 14,
    "historical-draft": 5,
    "historical-tutorial": 1,
}


class PapersCatalogError(RuntimeError):
    """The public catalog cannot be trusted or served."""


@dataclass(frozen=True)
class PapersCatalog:
    """Immutable exact-match allowlist derived from the public manifest."""

    slugs: frozenset[str]
    records: Mapping[str, Mapping[str, Any]]
    manifest_path: Path

    def contains(self, slug: str) -> bool:
        return bool(isinstance(slug, str) and SLUG_RE.fullmatch(slug) and slug in self.slugs)


def _positive_integer(value: Any, field: str) -> int:
    if not isinstance(value, int) or isinstance(value, bool) or value < 1:
        raise PapersCatalogError(f"invalid positive integer: {field}")
    return value


def _required_text(value: Any, field: str, max_length: int) -> str:
    if not isinstance(value, str) or not value.strip() or len(value) > max_length:
        raise PapersCatalogError(f"invalid text field: {field}")
    if any(ord(character) < 32 and character not in "\t\n\r" for character in value):
        raise PapersCatalogError(f"control character in field: {field}")
    return value.strip()


def _validate_source_url(value: Any) -> None:
    source = _required_text(value, "paper.source_url", 500)
    parsed = urlsplit(source)
    if (
        parsed.scheme != "https"
        or parsed.hostname != "github.com"
        or parsed.port is not None
        or parsed.username is not None
        or parsed.password is not None
        or parsed.query
        or parsed.fragment
        or not parsed.path.startswith(SOURCE_PREFIX)
    ):
        raise PapersCatalogError("paper source URL is outside the allowlist")


def _validate_meta(meta: Any) -> tuple[int, int, int, int]:
    if not isinstance(meta, dict):
        raise PapersCatalogError("manifest meta must be an object")
    if meta.get("schema_version") != SCHEMA_VERSION:
        raise PapersCatalogError("unsupported papers manifest schema")
    asset_version = meta.get("asset_version")
    if not isinstance(asset_version, str) or not ASSET_VERSION_RE.fullmatch(
        asset_version
    ):
        raise PapersCatalogError("invalid papers asset version")
    if meta.get("slug_pattern") != SLUG_PATTERN:
        raise PapersCatalogError("papers slug contract drift")
    contract = meta.get("result_contract")
    if not isinstance(contract, dict):
        raise PapersCatalogError("result boundary contract is missing")
    expected_contract = {
        "schema_version": "empirical-result-card-v1",
        "evidence_level": "historical_internal_record",
        "outcome_status": "not_normalized_in_current_ledger",
        "ledger_status": "not_bound",
        "review_status": "internal_only",
    }
    if any(contract.get(key) != expected for key, expected in expected_contract.items()):
        raise PapersCatalogError("result boundary contract drift")
    return (
        _positive_integer(meta.get("total_items"), "meta.total_items"),
        _positive_integer(meta.get("historical_result_records"), "meta.historical_result_records"),
        _positive_integer(meta.get("historical_research_drafts"), "meta.historical_research_drafts"),
        _positive_integer(meta.get("historical_tutorials"), "meta.historical_tutorials"),
    )


def _load_payload(manifest_path: Path) -> dict[str, Any]:
    try:
        value = json.loads(manifest_path.read_text(encoding="utf-8"))
    except (OSError, UnicodeError, json.JSONDecodeError) as exc:
        raise PapersCatalogError(f"cannot read papers manifest: {manifest_path}") from exc
    if not isinstance(value, dict):
        raise PapersCatalogError("papers manifest root must be an object")
    return value


@lru_cache(maxsize=8)
def load_papers_catalog(manifest_path: str | Path) -> PapersCatalog:
    """Load, cross-check, and cache the exact public slug allowlist."""

    path = Path(manifest_path).resolve()
    payload = _load_payload(path)
    total, result_count, draft_count, tutorial_count = _validate_meta(payload.get("meta"))
    groups = payload.get("groups")
    if not isinstance(groups, list) or len(groups) != len(GROUP_COUNTS):
        raise PapersCatalogError("papers manifest group count drift")

    records: dict[str, Mapping[str, Any]] = {}
    statuses = {status: 0 for status in STATUS_COUNTS}
    for index, (expected_group, expected_count) in enumerate(GROUP_COUNTS.items()):
        group = groups[index]
        if not isinstance(group, dict) or group.get("id") != expected_group:
            raise PapersCatalogError("papers manifest group order drift")
        papers = group.get("papers")
        if not isinstance(papers, list) or len(papers) != expected_count:
            raise PapersCatalogError(f"paper group count drift: {expected_group}")
        for paper in papers:
            if not isinstance(paper, dict):
                raise PapersCatalogError("paper record must be an object")
            slug = _required_text(paper.get("slug"), "paper.slug", 120)
            if not SLUG_RE.fullmatch(slug):
                raise PapersCatalogError(f"invalid paper slug: {slug}")
            if slug in records:
                raise PapersCatalogError(f"duplicate paper slug: {slug}")
            if "external_link" in paper:
                raise PapersCatalogError("paper detail routes must be internal")
            status = paper.get("status")
            if status not in statuses:
                raise PapersCatalogError(f"invalid paper status: {status}")
            _required_text(paper.get("title_zh"), "paper.title_zh", 300)
            _required_text(paper.get("title_en"), "paper.title_en", 300)
            _validate_source_url(paper.get("source_url"))
            statuses[status] += 1
            records[slug] = MappingProxyType(dict(paper))

    expected_counts = (20, 14, 5, 1)
    if (total, result_count, draft_count, tutorial_count) != expected_counts:
        raise PapersCatalogError("papers manifest count contract drift")
    if total != result_count + draft_count + tutorial_count or len(records) != total:
        raise PapersCatalogError("papers manifest total does not reconcile")
    if statuses != STATUS_COUNTS:
        raise PapersCatalogError("papers manifest status counts do not reconcile")

    markdown_dir = path.parent / "papers"
    markdown_slugs = frozenset(item.stem for item in markdown_dir.glob("*.md") if item.is_file())
    record_slugs = frozenset(records)
    if markdown_slugs != record_slugs:
        missing = sorted(record_slugs - markdown_slugs)
        orphaned = sorted(markdown_slugs - record_slugs)
        raise PapersCatalogError(
            f"manifest/Markdown slug drift; missing={missing}, orphaned={orphaned}"
        )
    return PapersCatalog(
        slugs=record_slugs,
        records=MappingProxyType(records),
        manifest_path=path,
    )
